File: src/app/views/donate_notifications.py
from datetime import datetime

from fastapi import APIRouter, Depends

from app.depends import check_token
from app.models import ChatDonateNotification


NOTIFICATION_DELTA_DAYS = 30


donation_notifications_router = APIRouter(
    prefix="/donate_notifications",
    tags=["donate_notifications"],
    dependencies=[Depends(check_token)],
)


@donation_notifications_router.get("/{chat_id}/is_need_send")
async def is_need_send(chat_id: int) -> bool:
    # add redis cache
    notification = await ChatDonateNotification.objects.get_or_none(chat_id=chat_id)

    if notification is None:
        return True

    delta = datetime.now() - notification.sended
    return delta.days >= NOTIFICATION_DELTA_DAYS


@donation_notifications_router.post("/{chat_id}")
async def mark_sended(chat_id: int):
    notification, created = await ChatDonateNotification.objects.get_or_create(
        _default={"sended": datetime.now()}, chat_id=chat_id
    )

    if created:
        return

    notification.sended = datetime.now()
    await notification.save()

File: src/views/donate_notifications.rs
use axum::{
    extract::{Path, Query},
    http::StatusCode,
    response::IntoResponse,
    routing::{get, post},
    Json, Router,
};
use chrono::Duration;
use serde::Deserialize;

use super::Database;

#[derive(sqlx::FromRow)]
struct ChatDonateNotification {
    pub sended: chrono::NaiveDateTime,
}

#[derive(Deserialize)]
struct IsNeedSendQuery {
    is_private: bool,
}

async fn is_need_send(
    Path(chat_id): Path<i64>,
    query: Query<IsNeedSendQuery>,
    db: Database,
) -> impl IntoResponse {
    const NOTIFICATION_DELTA_DAYS_PRIVATE: i64 = 60;
    const NOTIFICATION_DELTA_DAYS: i64 = 7;

    let notification = sqlx::query_as!(
        ChatDonateNotification,
        r#"SELECT sended FROM chat_donate_notifications WHERE chat_id = $1"#,
        chat_id
    )
    .fetch_optional(&db.0)
    .await
    .unwrap();

    let delta_days = if query.is_private {
        NOTIFICATION_DELTA_DAYS_PRIVATE
    } else {
        NOTIFICATION_DELTA_DAYS
    };

    match notification {
        Some(notification) => {
            let result = notification.sended + Duration::days(delta_days)
                <= chrono::offset::Local::now().naive_local();
            Json(result).into_response()
        }
        None => Json(true).into_response(),
    }
}

async fn mark_sent(Path(chat_id): Path<i64>, db: Database) -> impl IntoResponse {
    sqlx::query_as!(
        ChatDonateNotification,
        r#"INSERT INTO chat_donate_notifications (chat_id, sended) VALUES ($1, $2)
        ON CONFLICT (chat_id) DO UPDATE SET sended = EXCLUDED.sended
        RETURNING sended"#,
        chat_id,
        chrono::offset::Local::now().naive_local()
    )
    .fetch_one(&db.0)
    .await
    .unwrap();

    StatusCode::OK
}

pub fn get_router() -> Router {
    Router::new()
        .route("/:chat_id/is_need_send", get(is_need_send))
        .route("/:chat_id", post(mark_sent))
}
